### PR TITLE
refactor(*): implement a new design for interfaces

### PR DESCRIPTION
**Pull request by blackHatMonkey**
Implement a new design for interfaces module where getter and setters
are removed. Some of the methods have been removed or replaced to
make using network manager class much easier. Error messages are
simplified as well. Most importantly a tests suite is written for
the module and the coverage is 100%.

### CODE REVIEW

**Comment by sophron** [comment]
Nice, properties and unit-testing. I'll have a look and possibly merge soon.

File: wifiphisher/common/interfaces.py [hunk 1]
```diff
@@ -217,180 +272,220 @@ def __init__(self):
...
+        :param self: A NetworkManager object
+        :type self: NetworkManager
+        :return: None
+        :rtype: None
+        """
+
+        # populate our dictionary with all the available interfaces on the system
+        for interface in pyw.interfaces():
+            try:
+                card = pyw.getcard(interface)
```
Comment (sophron) [inline wifiphisher/common/interfaces.py]: This seems to throw a `pyric.error` (Errno 19: No such device)

The problem is that that `pyw.interfaces` returns all interfaces in the system including loopback but there's no device card tied to the loopback interface.

File: wifiphisher/common/interfaces.py [hunk 2]
```diff
@@ -1,163 +1,147 @@
...
 """
 
 import pyric
 import pyric.pyw as pyw
 
 
-class NotEnoughInterfacesFoundError(Exception):
-    """
-    Exception class to raise in case of a finding less than enough interfaces.
-    """
```
Comment (sophron) [inline wifiphisher/common/interfaces.py]: Why is that removed? I don't see where you check the number of available interfaces in the system in the new code.

File: wifiphisher/common/interfaces.py [hunk 3]
```diff
@@ -217,180 +272,220 @@ def __init__(self):
...
-            if interface_name == None:
-                if interface.has_ap_mode():
-                    return interface
-            if k == interface_name and not interface.being_used:
-                if interface.has_ap_mode():
-                    return interface
+        """
+
+        monitor_interface = self.get_interface(has_monitor_mode=True)
+        ap_interface = self.get_interface(has_ap_mode=True)
```
Comment (sophron) [inline wifiphisher/common/interfaces.py]: What is the problem with our current algorithm that automatically picks the right interfaces?

The new code doesn't look correct. Imagine the following scenario:

wlan0: supports monitor and ap mode
wlan1: supports monitor mode only

In that case, `ap_interface` should get wlan0 and `monitor_interface` should get `wlan1` but with this code they won't be assigned like that.

File: wifiphisher/pywifiphisher.py [hunk 4]
```diff
@@ -569,52 +575,55 @@ def start(self):
...
                 mon_iface = ap_iface
-                self.network_manager.set_ap_iface(ap_iface.get_name())
+
                 print ("[{0}+{1}] Selecting {0}{2}{1} interface for creating the "
-                       "rogue Access Point").format(G, W, ap_iface.get_name())
+                       "rogue Access Point").format(G, W, ap_iface)
+
+            # make sure interfaces are not blocked
+            self.network_manager.unblock_interface(ap_iface)
+            self.network_manager.unblock_interface(mon_iface)
```
Comment (sophron) [inline wifiphisher/pywifiphisher.py]: In various places, we also need to make sure that the cards are not invalid. See my comment on #580.

File: tests/test_interfaces.py [hunk 5]
```diff
@@ -0,0 +1,595 @@
...
+        """
+
+        interface_name = "wlan0"
+
+        with self.assertRaises(interfaces.InvalidInterfaceError):
+            self.network_manager.is_interface_valid(interface_name)
+
+    def test_is_interface_valid_no_ap_error(self):
+        """
+        Tests is_interface_valid method when interface is has no AP
```
Comment (sophron) [inline tests/test_interfaces.py]: * when interface has

File: tests/test_interfaces.py [hunk 6]
```diff
@@ -0,0 +1,595 @@
...
+        adapter = interfaces.NetworkAdapter(interface_name, interface_object)
+        adapter.has_ap_mode = False
+        self.network_manager._name_to_object[interface_name] = adapter
+
+        with self.assertRaises(interfaces.InvalidInterfaceError):
+            self.network_manager.is_interface_valid(interface_name, "AP")
+
+    def test_is_interface_valid_has_ap_none(self):
+        """
+        Tests is_interface_valid method when interface is has AP
```
Comment (sophron) [inline tests/test_interfaces.py]: * when interface has

This typo is in many places.

File: tests/test_interfaces.py [hunk 7]
```diff
@@ -0,0 +1,595 @@
...
+        self.network_manager._name_to_object[interface_name_1] = adapter_1
+
+        with self.assertRaises(interfaces.InterfaceCantBeFoundError):
+            self.network_manager.get_interface_automatically()
+
+    def test_get_interface_automatically_has_interface_interfaces(self):
+        """
+        Tests get_interface_automatically method when two interfaces
+        can be found
+        """
```
Comment (sophron) [inline tests/test_interfaces.py]: This is one of the most important tests. But you only examine one simple case.

You should examine all possible weird corner cases.

File: tests/test_interfaces.py [hunk 8]
```diff
@@ -0,0 +1,595 @@
+# pylint: skip-file
+""" This module tests the interface module """
```
Comment (sophron) [inline tests/test_interfaces.py]: I'm not sure I understand how you are asserting? I see some methods included from Pyric. Or how you do you mock the interfaces? Whatever you do, please document it.  :)

File: wifiphisher/common/interfaces.py (hunk 2 above)
Comment (blackHatMonkey) [inline wifiphisher/common/interfaces.py]: Checking for the number of interfaces in not needed as I demonstrate with this example:

Cards:
- AP

This will throw an error stating that an interface with monitor mode is required but no found.

Cards:
- AP
- MON

There is no problem and would continue as normal.

Cards:
- MON


This will throw an error stating that an interface with AP mode is required but no found.

File: wifiphisher/common/interfaces.py (hunk 3 above)
Comment (blackHatMonkey) [inline wifiphisher/common/interfaces.py]: I thought about it myself but it seems that I have messed up the order. Using your example swapping them should work since if interface has AP it is guaranteed to have monitor mode as well.

File: wifiphisher/common/interfaces.py (hunk 1 above)
Comment (blackHatMonkey) [inline wifiphisher/common/interfaces.py]: I didn't catch that error during my tests. I will add the exception.

File: tests/test_interfaces.py (hunk 8 above)
Comment (blackHatMonkey) [inline tests/test_interfaces.py]: I don't understand what you mean. I just check that the code calls `PyRIC`'s function and in some cases change the return values for some functions.

File: tests/test_interfaces.py (hunk 6 above)
Comment (blackHatMonkey) [inline tests/test_interfaces.py]: Thats what you get for copy-pasting :cry: .

File: tests/test_interfaces.py (hunk 7 above)
Comment (blackHatMonkey) [inline tests/test_interfaces.py]: I will add more tests in there.

File: wifiphisher/common/interfaces.py (hunk 3 above)
Comment (sophron) [inline wifiphisher/common/interfaces.py]: >  if interface has AP it is guaranteed to have monitor mode as well

I'm not sure that's correct. What's your source?

File: wifiphisher/common/interfaces.py (hunk 2 above)
Comment (sophron) [inline wifiphisher/common/interfaces.py]: I think this will raise confusion. If a user has one card that supports both AP and Monitor modes, he will see an error that he needs an interface that supports Monitor Mode. It won't be clear to the user that he needs a separate card for Monitor Mode.

We need to be clear when raising an exception. Not having enough adapters is different from having the required number of adapters but they are not supporting the right modes. Otherwise, people will start opening issues about what is required again.

Comment (blackHatMonkey) [inline wifiphisher/common/interfaces.py]: Imagine this scenario where I have a wired interface and a wireless interface. It passes the two interface test while throwing the same error.

Comment (sophron) [inline wifiphisher/common/interfaces.py]: Of'course the check would be if there are two wireless interfaces (`pyric.iswireless`) installed.

Comment (blackHatMonkey) [inline wifiphisher/common/interfaces.py]: I think the best option would be to reword this to include a message regarding the number of interfaces.

File: wifiphisher/common/interfaces.py (hunk 3 above)
Comment (blackHatMonkey) [inline wifiphisher/common/interfaces.py]: I will try to come up with a better algorithm to fix this.

**Comment by blackHatMonkey** [comment]
@sophron @anakin1028 I have made some changes to this PR. Can you please review it and let me know if any changes need to be made.

Disclaimer: I haven't had the chance to test it so it might not work.

File: wifiphisher/common/interfaces.py [hunk 9]
```diff
@@ -319,233 +339,320 @@ def __init__(self):
...
+        :return: None
+        :rtype: None
+        .. note: This method will set the interface to managed mode
+        """
+
+        card = self._name_to_object[interface_name].card
+        self.set_interface_mode(interface_name, "managed")
+
+        try:
+            pyw.macset(card, mac_address)
```
Comment (anakin1028) [inline wifiphisher/common/interfaces.py]: Before setting mac address we should check whether the card is down. This is my first error to executing on this branch. I will fixed it on my local and see if there are any problems else.

File: wifiphisher/pywifiphisher.py [hunk 10]
```diff
@@ -586,69 +590,73 @@ def start(self):
...
 
         if not args.no_mac_randomization:
-            print ("[{0}+{1}] " + ap_iface.get_name() + ' mac address becomes '
-                     + ap_iface.get_current_mac()).format(G, W)
+            print "[{0}+{1}] {2} mac address becomes is now {3} ".format(G, W, ap_iface)
+
             if not args.nojamming:
-                print ("[{0}+{1}] " + mon_iface.get_name() + ' mac address becomes '
-                     + mon_iface.get_current_mac()).format(G, W)
+                print ("[{0}+{1}] {2} mac address becomes {3}".format(G, W, mon_iface))
```
Comment (anakin1028) [inline wifiphisher/pywifiphisher.py]: you don't give the mac address argument

Comment (anakin1028) [inline wifiphisher/pywifiphisher.py]: line#669, `up_ifaces` is not defined in `NetworkManager`

File: wifiphisher/common/interfaces.py (hunk 9 above)
Comment (blackHatMonkey) [inline wifiphisher/common/interfaces.py]: is `pyw.up(card)` enough?

**Comment by blackHatMonkey** [comment]
@anakin1028 How did you see the mac randomization? Do you think we can do it better?

File: wifiphisher/pywifiphisher.py (hunk 10 above)
Comment (anakin1028) [inline wifiphisher/pywifiphisher.py]: in `select_access_point` you should pass the `NetworkManager` inside.

Comment (anakin1028) [inline wifiphisher/pywifiphisher.py]: check this statement `self.access_point.set_interface(ap_iface.get_name())`. `ap_iface` is already a string.

**Comment by anakin1028** [comment]
Brian, after the problems I have mentioned are fixed, I'll do the functional test :)

File: wifiphisher/common/interfaces.py (hunk 9 above)
Comment (anakin1028) [inline wifiphisher/common/interfaces.py]: also checks if the card is in `managed` mode

File: wifiphisher/common/interfaces.py [hunk 11]
```diff
@@ -319,233 +339,320 @@ def __init__(self):
...
         :rtype: None
-        .. seealso:: NetworkAdapter
         """
 
-        # Setup the fields
-        self._interfaces = {}
-        self.ap_iface = ""
-        self.jam_iface = ""
+        self._name_to_object = dict()
+        self._active = set()
```
Comment (anakin1028) [inline wifiphisher/common/interfaces.py]: is the name `_name_to_object` clear enough to represent it is a dictionary doing mapping from adapter name to the `NetworkAdapter`?

File: wifiphisher/common/interfaces.py [hunk 12]
```diff
@@ -319,233 +339,320 @@ def __init__(self):
...
+            if mode == "monitor" and not interface_adapter.has_monitor_mode:
+                raise InvalidInterfaceError(interface_name, mode)
+            elif mode == "AP" and not interface_adapter.has_ap_mode:
+                raise InvalidInterfaceError(interface_name, mode)
 
-    def up_ifaces(self, ifaces):
-        for i in ifaces:
-            card = pyw.getcard(i.get_name())
-            pyw.up(card)
-
```
Comment (anakin1028) [inline wifiphisher/common/interfaces.py]: why `up_ifaces` removed? I remember in `pywifiphisher.py` will call this function, or you have refactored this part too?

File: tests/test_interfaces.py [hunk 13]
```diff
@@ -0,0 +1,779 @@
...
+
+import unittest
+import mock
+import wifiphisher.common.interfaces as interfaces
+import pyric
+
```
Comment (sophron) [inline tests/test_interfaces.py]: You will have to set your path accordingly to execute the tests from whatever location. See how I did it in my tests.

File: tests/test_interfaces.py [hunk 14]
```diff
@@ -0,0 +1,779 @@
...
+
+class TestNetworkAdapter(unittest.TestCase):
+    """ Tests NetworkAdapter class """
+
+    @mock.patch("wifiphisher.common.interfaces.pyw")
```
Comment (sophron) [inline tests/test_interfaces.py]: What is the reason for mocking here?

File: tests/test_interfaces.py [hunk 15]
```diff
@@ -0,0 +1,779 @@
...
+        Tests is_interface_wired when interface is wired
+        """
+
+        interface_name = "lan0"
+        interface_object = "Card Object"
+        adapter = interfaces.NetworkAdapter(interface_name, interface_object, self.mac_address)
+        adapter.is_wireless = False
+        self.network_manager._name_to_object[interface_name] = adapter
+
+        self.network_manager.is_interface_wired(interface_name)
```
Comment (sophron) [inline tests/test_interfaces.py]: I don't see any `assert` here.

**Comment by sophron** [comment]
Brian,

A few notes on your tests.

In general, unit tests should check the behavior of the code that you have written. The test `test_has_monitor_mode_true` that checks if a property of an object has been set properly to True, is a unit test in Python implementation itself and not in your code. If, however, there was a logic that you have written behind setting the property (e.g. the monitor interface would be true in specific cases) then it would make sense to write a test for it.

Same applies for most of your tests. Tests that check if the properties have been set properly (e.g. `test_mac_address_original` and `test_mac_address_modified`) are not offering something. A test however that would check the `recover_mac_address_to_original` method if it did recover the mac address to its previous state would make sense.

All in all, what we want to test is the logic of the tool, not if Python sets the properties correctly. The first case would break if the application logic changes, the second type would break simply if we rename a field. It's important to have specific, not-useless test cases that we can easily maintain and people would get the idea of a module simply by reading them.

File: tests/test_interfaces.py (hunk 13 above)
Comment (blackHatMonkey) [inline tests/test_interfaces.py]: instead of doing that for all of our tests we should use nose or some similar tool. All you have to do is run nose from the `wifiphisher` directory.

File: wifiphisher/common/interfaces.py (hunk 11 above)
Comment (blackHatMonkey) [inline wifiphisher/common/interfaces.py]: do you have something else in mind?

File: wifiphisher/common/interfaces.py (hunk 9 above)
Comment (blackHatMonkey) [inline wifiphisher/common/interfaces.py]: >Before setting mac address we should check whether the card is down. This is my first error to executing on this branch. I will fixed it on my local and see if there are any problems else.

Since I'm setting the card to `managed on line 395 the interface should b e 100% up.

>also checks if the card is in managed mode

On line 395 I'm setting it to `managed`.

Comment (anakin1028) [inline wifiphisher/common/interfaces.py]: @blackHatMonkey , yes. But mac address can only change when the card is in down state.

Comment (blackHatMonkey) [inline wifiphisher/common/interfaces.py]: Did not realize this. I will make the changes and document it :smile.

File: wifiphisher/common/interfaces.py (hunk 12 above)
Comment (blackHatMonkey) [inline wifiphisher/common/interfaces.py]: yes we no longer need this function. I just forgot to remove it :cry: .

File: tests/test_interfaces.py (hunk 14 above)
Comment (blackHatMonkey) [inline tests/test_interfaces.py]: I believe it was because of `pyw.card`. However we don't need it and I will remove it.

File: wifiphisher/common/interfaces.py (hunk 9 above)
Comment (anakin1028) [inline wifiphisher/common/interfaces.py]: OK cheers :)

File: tests/test_interfaces.py (hunk 13 above)
Comment (blackHatMonkey) [inline tests/test_interfaces.py]: Or maybe we can leverage `tox`?

**Comment by blackHatMonkey** [comment]
@sophron @anakin1028 I have fixed all the problems you mentioned. I have done a basic test and it should work now but it requires a full testing.

@sophron about the `test_has_monitor_mode_true` I'm checking the property and not the actual variable. So even if we change the variable it should still work. Also I have fixed a bunch other cases where I was checking for variables :smile:.

**Comment by blackHatMonkey** [comment]
@anakin1028 By `doing better` I meant that right now we are doing randomization in two places. I think we should come up with a better plan so that we only have to do it once.

**Comment by sophron** [comment]
Nice stuff!

@anakin1028 feel free to merge this if you have no other comments.

**Comment by anakin1028** [comment]
@blackHatMonkey Nice~~~ I just help fixed one small problem, that is in `pywifiphisher` the `mon_iface` variable is string, but you still use it to set the channel number. otherwise it looks good, I'll merge it to master :)
